File: web/src/mocks/fixtures/cacheEntries.ts
// Cache entries for the browse UI.
//
// Deliberately not uniform. The states below all exist in a real store and each
// one is a rendering decision the page has to get right, so the fixture carries
// them rather than leaving them to be discovered against a live server:
//
//   * entries written before 0.5, which have a `provider_fingerprint` and no
//     `request` at all (the drawer must fall back rather than show nothing),
//   * entries the backfill has not reached (`indexed: false` — "indexing", not
//     "no model"),
//   * a body this server could not parse (`parseable: false`),
//   * a ≤0.4.x grader entry, whose kind is inferred from its verdict tag,
//   * one very large output, which the drawer must refuse to auto-expand,
//   * one deeply nested `raw`, which must not go through the JSON tree,
//   * missing usage and missing cost, which must render as "-" rather than 0,
//   * entries whose output came back empty and why (`empty_reason`), which is
//     what the reason filter and its facet exist to find — a store with none of
//     them would let an empty dropdown pass for a working one.

import type {
  CacheEntryDetail,
  CacheEntryListItem,
  CacheEntryRunsResponse,
  CacheFacetsResponse,
} from "@/api";
import { hash, rand } from "./rng";

const NOW = Date.parse("2026-07-28T12:00:00Z");
const MINUTE = 60_000;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;

/** Above the drawer's auto-expand threshold, so the guard is exercised. */
const HUGE_OUTPUT_BYTES = 320_000;

const MODELS = [
  "claude-opus-5",
  "claude-sonnet-5",
  "gpt-4o",
  "text-embedding-3-large",
] as const;

const KINDS = ["provider", "judge", "embedding", "exec_assert"] as const;

function toIso(ms: number): string {
  return new Date(ms).toISOString();
}

/** A stable, well-formed `sha256:` key for an index. */
function keyFor(i: number): string {
  let hex = "";
  for (let block = 0; block < 8; block++) {
    hex += hash("cache-entry", i, block).toString(16).padStart(8, "0");
  }
  return `sha256:${hex.slice(0, 64)}`;
}

interface Shape {
  /** Not yet reached by the backfill: metadata is unknown, not absent. */
  unindexed?: boolean;
  /** Examined, and the body was not something this server understands. */
  unparseable?: boolean;
  /** Written before 0.5: a fingerprint instead of a request. */
  preRequest?: boolean;
  /** A ≤0.4.x grader entry. */
  legacyVerdict?: boolean;
  huge?: boolean;
  deepRaw?: boolean;
  noUsage?: boolean;
  /** The output came back empty, and the entry records why. */
  emptyReason?: string;
}

/**
 * Which of the 180 entries deviate from the ordinary shape.
 *
 * Three distinct `empty_reason` values with different counts, deliberately: a
 * facet dropdown with one option cannot show whether it sorts by count, and a
 * single reason would let the page look right while `?empty_reason=` was
 * hard-coded to it.
 */
function shapeOf(i: number): Shape {
  if (i < 12) return { preRequest: true };
  if (i < 32) return { unindexed: true };
  if (i === 32) return { unparseable: true };
  if (i === 33) return { huge: true };
  if (i === 34) return { deepRaw: true };
  if (i < 37) return { legacyVerdict: true };
  if (i < 40) return { noUsage: true };
  // Eleven poisoned entries, weighted toward `refusal` the way a real store is.
  if (i < 46) return { emptyReason: "refusal" };
  if (i < 50) return { emptyReason: "blank" };
  if (i === 50) return { emptyReason: "truncated" };
  return {};
}

function modelFor(i: number): string {
  return MODELS[Math.floor(rand("model", i) * MODELS.length)] as string;
}

function kindFor(i: number, shape: Shape): string {
  if (shape.legacyVerdict) return "judge";
  const model = modelFor(i);
  if (model.startsWith("text-embedding")) return "embedding";
  return KINDS[Math.floor(rand("kind", i) * KINDS.length)] as string;
}

function requestSummaryFor(kind: string): string {
  if (kind === "exec_assert") return "exec ./graders/schema-check";
  if (kind === "embedding") return "POST https://api.openai.com/v1/embeddings";
  return "POST https://api.anthropic.com/v1/messages";
}

const OUTPUTS = [
  "Your refund window is thirty days from delivery.",
  "I can help with that. First, confirm the order number.",
  "The account is locked after five failed attempts.",
  "Escalating to a human agent — one moment.",
  "That plan includes unlimited seats and priority support.",
] as const;

function outputFor(i: number, shape: Shape): string {
  if (shape.huge) return "x".repeat(HUGE_OUTPUT_BYTES);
  // An `empty_reason` is only ever computed for a blank output, so a fixture
  // that carried one beside prose would describe a state the server cannot
  // produce — and would hide the "nothing to show, here is why" rendering.
  if (shape.emptyReason) return "";
  return OUTPUTS[Math.floor(rand("output", i) * OUTPUTS.length)] as string;
}

/** A `raw` payload wide enough that a naive JSON tree would stall on it. */
function deepRaw(): unknown {
  const content: Record<string, unknown> = {};
  for (let i = 0; i < 2000; i++) {
    content[`token_${i}`] = { logprob: -0.0001 * i, rank: i };
  }
  return { id: "msg_deep", model: "claude-opus-5", logprobs: content };
}

export function cacheEntryList(): CacheEntryListItem[] {
  return Array.from({ length: 180 }, (_, i) => {
    const shape = shapeOf(i);
    const created = NOW - Math.floor(rand("created", i) * 40 * DAY);
    const size = shape.huge
      ? HUGE_OUTPUT_BYTES + 512
      : 900 + Math.floor(rand("size", i) * 14_000);

    if (shape.unindexed) {
      return {
        key: keyFor(i),
        size,
        created_at: toIso(created),
        last_access_at: toIso(created + HOUR),
        entry_created_at: null,
        indexed: false,
        parseable: null,
        kind: null,
        model: null,
        cost_usd: null,
        input_tokens: null,
        output_tokens: null,
        request_summary: null,
        output_preview: null,
        empty_reason: null,
      };
    }
    if (shape.unparseable) {
      return {
        key: keyFor(i),
        size,
        created_at: toIso(created),
        last_access_at: toIso(created + HOUR),
        entry_created_at: null,
        indexed: true,
        parseable: false,
        kind: null,
        model: null,
        cost_usd: null,
        input_tokens: null,
        output_tokens: null,
        request_summary: null,
        output_preview: null,
        // Nothing was parsed, so nothing is known about why an output was
        // empty. `parseable: false` is the field that says so.
        empty_reason: null,
      };
    }

    const kind = kindFor(i, shape);
    const inputTokens = shape.noUsage ? null : 200 + Math.floor(rand("in", i) * 3000);
    const outputTokens = shape.noUsage ? null : 40 + Math.floor(rand("out", i) * 900);
    return {
      key: keyFor(i),
      size,
      created_at: toIso(created),
      last_access_at: toIso(created + Math.floor(rand("access", i) * 3 * DAY)),
      entry_created_at: toIso(created - MINUTE),
      indexed: true,
      parseable: true,
      kind,
      model: shape.preRequest ? null : modelFor(i),
      cost_usd:
        shape.noUsage || kind === "embedding"
          ? null
          : Number((rand("cost", i) * 0.09).toFixed(6)),
      input_tokens: inputTokens,
      output_tokens: outputTokens,
      request_summary: shape.preRequest ? null : requestSummaryFor(kind),
      output_preview: outputFor(i, shape).slice(0, 120),
      empty_reason: shape.emptyReason ?? null,
    } satisfies CacheEntryListItem;
  });
}

/** The detail view of one entry, keyed by its `sha256:` string. */
export function cacheEntryDetail(
  key: string,
  includeRaw: boolean,
): CacheEntryDetail | null {
  const index = cacheEntryList().findIndex((e) => e.key === key);
  if (index < 0) return null;
  const row = cacheEntryList()[index] as CacheEntryListItem;
  const shape = shapeOf(index);

  const base: CacheEntryDetail = {
    key: row.key,
    size: row.size,
    created_at: row.created_at,
    last_access_at: row.last_access_at,
    entry_created_at: row.entry_created_at,
    indexed: row.indexed,
    parseable: row.parseable,
    kind: row.kind,
    model: row.model,
    cost_usd: row.cost_usd,
    input_tokens: row.input_tokens,
    output_tokens: row.output_tokens,
    attempts: null,
    provider_latency_ms: null,
    stop_reason: null,
    empty_reason: row.empty_reason,
    domarinn_version: null,
    request: null,
    provider_fingerprint: null,
    output: null,
    reasoning: null,
    tool_calls: [],
    raw: null,
  };

  if (shape.unindexed || shape.unparseable) return base;

  const kind = row.kind ?? "provider";
  return {
    ...base,
    attempts: kind === "provider" ? 1 : null,
    provider_latency_ms: 400 + Math.floor(rand("latency", index) * 2600),
    stop_reason: kind === "provider" ? "stop" : null,
    domarinn_version: shape.preRequest ? "0.4.2" : "0.5.0",
    // Pre-0.5 entries carry one or the other, never both: the fingerprint
    // stopped being written once the request itself was recorded.
    request: shape.preRequest
      ? null
      : kind === "exec_assert"
        ? {
            transport: "exec",
            command: "./graders/schema-check",
            args: ["--strict"],
            stdin: { domarinn: { protocol: 1, kind: "assert" } },
          }
        : {
            transport: "http",
            method: "POST",
            url: "https://api.anthropic.com/v1/messages",
            body: {
              model: row.model,
              max_tokens: 1024,
              messages: [
                { role: "user", content: "How long is the refund window?" },
              ],
            },
          },
    provider_fingerprint: shape.preRequest
      ? { type: "anthropic", model: "claude-3-5-sonnet" }
      : null,
    output: outputFor(index, shape),
    reasoning:
      kind === "provider" && rand("reasoning", index) > 0.7
        ? "Checked the policy table, then the exceptions list."
        : null,
    tool_calls: [],
    raw: includeRaw
      ? shape.deepRaw
        ? deepRaw()
        : { id: `msg_${index}`, role: "assistant", stop_reason: "end_turn" }
      : null,
  };
}

/**
 * Runs that used an entry.
 *
 * Most entries link to nothing, which is the honest common case: the key is
 * only recorded by runs from a version that knew to write one, and older runs
 * cannot be backfilled.
 */
export function cacheEntryRuns(key: string): CacheEntryRunsResponse {
  const index = cacheEntryList().findIndex((e) => e.key === key);
  if (index < 0 || index % 3 !== 0) return { cases: [], next_cursor: null };
  const count = 1 + (index % 3);
  return {
    cases: Array.from({ length: count }, (_, i) => ({
      run_id: `checkout-agent-regression-${12 - i}`,
      project: "checkout-agent",
      suite: "regression",
      created_at: toIso(NOW - (i + 1) * DAY),
      case_key: `case-${index}-${i}`,
      name: `refund policy #${index}`,
      status: i === 0 ? "pass" : "fail",
      cached: i > 0,
    })),
    next_cursor: null,
  };
}

export function cacheFacets(): CacheFacetsResponse {
  const rows = cacheEntryList();
  const count = (pick: (r: CacheEntryListItem) => string | null) => {
    const tally = new Map<string, number>();
    for (const row of rows) {
      const value = pick(row);
      if (value) tally.set(value, (tally.get(value) ?? 0) + 1);
    }
    return [...tally]
      .map(([value, n]) => ({ value, count: n }))
      .sort((a, b) => b.count - a.count);
  };
  return {
    kinds: count((r) => r.kind),
    models: count((r) => r.model),
    empty_reasons: count((r) => r.empty_reason),
    total: rows.length,
    unindexed: rows.filter((r) => !r.indexed).length,
    unparseable: rows.filter((r) => r.parseable === false).length,
  };
}

File: crates/domarinn-server/src/dto/cacheentries.rs
//! DTOs for browsing the cache: `GET /cache/entries`, `/cache/entries/{key}`
//! and `/cache/facets`.
//!
//! Separate from [`super::cache`], which serves the client-facing stats and
//! prune surface. These describe what an entry *is*, not how the cache is
//! doing.

use serde::Serialize;
use serde_json::Value as Json;
use ts_rs::TS;

use domarinn_core::result::ToolCall;
use domarinn_core::types::Output;

/// One row of the entries table.
///
/// Everything below `last_access_at` is derived from the entry body, so all of
/// it is nullable twice over: the body may not have carried the field, and the
/// body may not have been examined yet. `indexed` and `parseable` are how a
/// client tells those two apart instead of rendering both as "unknown".
#[derive(Debug, Clone, Serialize, TS)]
pub struct CacheEntryListItem {
    pub key: String,
    pub size: i64,
    /// When this instance was handed the entry, RFC3339.
    pub created_at: String,
    /// RFC3339. `None` on a tier that cannot report access times honestly.
    pub last_access_at: Option<String>,
    /// When the *call* happened, as the entry records it, RFC3339. Differs from
    /// `created_at` by however long the entry sat on a laptop before upload.
    pub entry_created_at: Option<String>,
    /// Whether the body has been examined yet. `false` means the columns below
    /// are unknown rather than absent — say "indexing", not "no model".
    pub indexed: bool,
    /// Whether examining it found an entry this server understands. `None`
    /// while `indexed` is false.
    pub parseable: Option<bool>,
    pub kind: Option<String>,
    pub model: Option<String>,
    pub cost_usd: Option<f64>,
    pub input_tokens: Option<i64>,
    pub output_tokens: Option<i64>,
    /// Where the request went — transport plus host or command, never prompt
    /// text and never a query string. See `storage::cacheindex`.
    pub request_summary: Option<String>,
    pub output_preview: Option<String>,
    /// Why the output was empty, when it was. A plain `Option` with **no**
    /// `skip_serializing_if`, unlike `dto::cases` and `dto::runs`, which omit
    /// theirs: this module's rule is that an absent member is an explicit
    /// `null` (pinned by `an_unindexed_row_serializes_every_unknown_as_explicit_null`),
    /// because a client here has to distinguish "not examined" from "examined,
    /// nothing to report" — which is also why the column is plain `NULL` rather
    /// than the runs side's `''` sentinel.
    pub empty_reason: Option<String>,
}

/// `GET /cache/entries` response.
#[derive(Debug, Clone, Serialize, TS)]
pub struct CacheEntryListResponse {
    pub entries: Vec<CacheEntryListItem>,
    pub next_cursor: Option<String>,
    /// True when the tier stopped looking before it ran out of entries. Only a
    /// scanning tier can set this; the server tier answers from an index and is
    /// always exhaustive.
    pub truncated: bool,
}

/// `GET /cache/entries/{key}` response.
///
/// Every body-derived member is `None` when `parseable` is false, which is a
/// real state rather than an error: the cache stores what it is handed, so an
/// entry written by a newer domarinn is served intact by `GET /cache/{key}`
/// and simply cannot be described here.
#[derive(Debug, Clone, Serialize, TS)]
pub struct CacheEntryDetail {
    pub key: String,
    pub size: i64,
    pub created_at: String,
    pub last_access_at: Option<String>,
    pub entry_created_at: Option<String>,
    pub indexed: bool,
    pub parseable: Option<bool>,
    pub kind: Option<String>,
    pub model: Option<String>,
    pub cost_usd: Option<f64>,
    pub input_tokens: Option<i64>,
    pub output_tokens: Option<i64>,
    pub attempts: Option<u32>,
    /// The original call's in-flight time, not the cache read's.
    pub provider_latency_ms: Option<u64>,
    pub stop_reason: Option<String>,
    /// Why the output was empty. See [`CacheEntryListItem::empty_reason`] for
    /// why this is an explicit `null` rather than an omitted key.
    pub empty_reason: Option<String>,
    pub domarinn_version: Option<String>,
    /// The redacted canonical request. Credentials live in headers and are
    /// structurally absent from both envelope shapes.
    #[ts(type = "unknown")]
    pub request: Option<Json>,
    /// What selected the provider that answered. Only on entries written before
    /// the request itself was recorded — the two are never both present.
    #[ts(type = "unknown")]
    pub provider_fingerprint: Option<Json>,
    pub output: Option<Output>,
    pub reasoning: Option<String>,
    pub tool_calls: Vec<ToolCall>,
    /// Raw provider metadata. Withheld unless `?raw=true`: it is the largest
    /// member and the least often wanted, and sending it on every drawer open
    /// would cost the whole payload a second time.
    #[ts(type = "unknown")]
    pub raw: Option<Json>,
}

/// One case that used an entry.
#[derive(Debug, Clone, Serialize, TS)]
pub struct CacheEntryRunRef {
    pub run_id: String,
    pub project: Option<String>,
    pub suite: Option<String>,
    pub created_at: String,
    pub case_key: String,
    pub name: Option<String>,
    pub status: String,
    /// Whether this particular case was served from cache. A miss that *wrote*
    /// the entry is listed too — the key belongs to the request, so both are
    /// runs that used this entry.
    pub cached: bool,
}

/// `GET /cache/entries/{key}/runs` response.
///
/// An empty list is ambiguous and readers must say so: cross-linking only works
/// for runs recorded by a version that wrote the key, and no backfill can
/// supply it for older ones. "No runs" is not evidence that an entry is unused.
#[derive(Debug, Clone, Serialize, TS)]
pub struct CacheEntryRunsResponse {
    pub cases: Vec<CacheEntryRunRef>,
    pub next_cursor: Option<String>,
}

/// One value a filter can take, and how many entries carry it.
#[derive(Debug, Clone, Serialize, TS)]
pub struct CacheFacet {
    pub value: String,
    pub count: i64,
}

/// `GET /cache/facets` response — what to put in the filter dropdowns.
#[derive(Debug, Clone, Serialize, TS)]
pub struct CacheFacetsResponse {
    pub kinds: Vec<CacheFacet>,
    /// Capped at the most common values: a store polluted with pathological
    /// model strings must not be able to produce an unbounded response.
    pub models: Vec<CacheFacet>,
    /// Deliberately uncapped, unlike `models`: the reason vocabulary is a
    /// handful of constants, each value is length-clamped on the way into the
    /// index, and this is the dropdown someone reaches for when a refusal has
    /// poisoned a shared cache — truncating its tail would hide the rare
    /// reason they are hunting.
    pub empty_reasons: Vec<CacheFacet>,
    pub total: i64,
    /// Entries whose body has not been examined yet. They are listed, but no
    /// `kind`/`model` filter can match them and search cannot reach them.
    pub unindexed: i64,
    /// Entries whose body this server could not parse.
    pub unparseable: i64,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn item() -> CacheEntryListItem {
        CacheEntryListItem {
            key: "sha256:aa".into(),
            size: 12,
            created_at: "2026-01-01T00:00:00+00:00".into(),
            last_access_at: Some("2026-01-02T00:00:00+00:00".into()),
            entry_created_at: None,
            indexed: true,
            parseable: Some(true),
            kind: Some("provider".into()),
            model: Some("claude-opus-5".into()),
            cost_usd: Some(0.5),
            input_tokens: Some(10),
            output_tokens: Some(20),
            request_summary: Some("POST https://api/x".into()),
            output_preview: Some("hi".into()),
            empty_reason: None,
        }
    }

    #[test]
    fn list_item_matches_todays_wire_shape() {
        assert_eq!(
            serde_json::to_value(item()).unwrap(),
            json!({
                "key": "sha256:aa",
                "size": 12,
                "created_at": "2026-01-01T00:00:00+00:00",
                "last_access_at": "2026-01-02T00:00:00+00:00",
                "entry_created_at": null,
                "indexed": true,
                "parseable": true,
                "kind": "provider",
                "model": "claude-opus-5",
                "cost_usd": 0.5,
                "input_tokens": 10,
                "output_tokens": 20,
                "request_summary": "POST https://api/x",
                "output_preview": "hi",
                "empty_reason": null,
            })
        );
    }

    /// Absent members are `null`, never omitted — the whole DTO layer's rule,
    /// and what lets the UI distinguish "no value" from "field not in this
    /// server's version".
    #[test]
    fn an_unindexed_row_serializes_every_unknown_as_explicit_null() {
        let mut row = item();
        row.indexed = false;
        row.parseable = None;
        row.kind = None;
        row.model = None;
        row.cost_usd = None;
        row.input_tokens = None;
        row.output_tokens = None;
        row.request_summary = None;
        row.output_preview = None;
        row.empty_reason = None;

        let v = serde_json::to_value(row).unwrap();
        for field in [
            "parseable",
            "kind",
            "model",
            "cost_usd",
            "input_tokens",
            "output_tokens",
            "request_summary",
            "output_preview",
            "empty_reason",
        ] {
            assert!(v.get(field).is_some(), "{field} was omitted");
            assert!(v[field].is_null(), "{field} was not null");
        }
    }

    #[test]
    fn an_exhausted_list_reports_a_null_cursor() {
        let v = serde_json::to_value(CacheEntryListResponse {
            entries: vec![],
            next_cursor: None,
            truncated: false,
        })
        .unwrap();
        assert!(v.get("next_cursor").is_some());
        assert!(v["next_cursor"].is_null());
        assert_eq!(v["truncated"], json!(false));
    }
}
